common/survey/_internal/surveyRefDataIndex.js

```diff
@@ -0,0 +1,160 @@
+const R = require('ramda')
+
+const ObjectUtils = require('../../../common/objectUtils')
+
+const SurveyNodeDefs = require('./surveyNodeDefs')
+const NodeDef = require('../../../common/survey/nodeDef')
+const RecordReader = require('../../../common/record/_internal/recordReader')
+const Node = require('../../../common/record/node')
+
+const CategoryItem = require('../../../common/survey/categoryItem')
+const CategoryLevel = require('../../../common/survey/categoryLevel')
+const Taxon = require('../../../common/survey/taxon')
+
+/**
+ * categoryItemUuidIndex : {
+ *    [$categoryUuid] : {
+ *      [$parentCategoryItemUuid] :{
+ *        [$categoryItemCode] : $categoryItemUuid
+ *      }
+ *    }
+ * }
+ *
+ * categoryItemIndex : {
+ *    [$categoryItemUuid] : { ...$categoryItem }
+ * }
+ *
+ * taxonUuidIndex : {
+ *   [$taxonomyUuid] : {
+ *     [$taxonCode] : {
+ *       uuid : $taxonUuid,
+ *       vernacularNames: {
+ *        [$vernacularName] : $vernacularNameUuid,
+ *       ...
+ *       }
+ *     }
+ *   }
+ * }
+ *
+ *  taxonIndex : {
+ *    [$taxonUuid] : { ...$taxon }
+ * }
+ *
+ */
+
+const keys = {
+  // root path key
+  indexRefData: '_indexRefData',
+  // ref data indexes
+  categoryItemUuidIndex: 'categoryItemUuidIndex',
+  categoryItemIndex: 'categoryItemIndex',
+  taxonUuidIndex: 'taxonUuidIndex',
+  taxonIndex: 'taxonIndex'
+}
+
+// ====== READ
+
+// ==== category index
+const getCategoryItemUuidAndCodeHierarchy = (survey, nodeDef, record, parentNode, code) => surveyIndex => {
```

Comment:
Why not split this into two separate functions? (getCategoryItemUuid and getCodeHierarchy)?

Reply:
to avoid duplicating the call to ```RecordReader.getParentCodeAttriubute``` which could be time consuming. since the results are used only together from outside, we thought to do once only the heavy computational logic